src/main/java/org/truffleruby/parser/lexer/HeredocTerm.java

```diff
@@ -103,12 +104,22 @@ public int parseString(RubyLexer lexer) {
             return error(lexer, eos);
         }
 
+        boolean bol = lexer.was_bol();
         // Found end marker for this heredoc
-        if (lexer.was_bol() && lexer.whole_match_p(nd_lit, indent)) {
-            lexer.heredoc_restore(this);
-            lexer.setStrTerm(null);
-            lexer.setState(EXPR_END);
-            return RubyParser.tSTRING_END;
+        if (bol) {
+            // if line_indent == -1 then we are after an interpolation in same line or there was a line continuation.
+            int line_indent = lexer.heredoc_line_indent;
+
+            if (line_indent != -1) {
+                if (lexer.whole_match_p(nd_lit, indent)) {
+                    lexer.heredoc_restore(this);
+                    lexer.setStrTerm(null);
+                    lexer.setState(EXPR_END);
+                    return RubyParser.tSTRING_END;
+                }
+            } else {
+                lexer.setHeredocLineIndent(0);
```

Comment:
So here is what is going on: `heredoc_line_indent` tracks the indentation on the current line, which is used to set `heredoc_indent` to the smallest indent among the lines in the heredoc string. The comments I've added to the code should help clarify this.

Resetting `heredoc_line_indent` to 0 is necessary because it's set to -1 after seen the first non-whitespace character on a line, and needs to be reset to 0 before processing the next line. (There's certainly a much better way to do this, but that's how it works currently.)